### ISSUE TITLE
Extend std.LinkedList with single/tailq and circularq variants

### ISSUE THREAD

**Issue by daurnimator** [ziglang/zig#2384]
The [classic `sys/queue.h`](https://linux.die.net/man/3/circleq_entry) has 3 variants of linked list: `LIST`, `TAILQ`, `CIRCULARQ`.

> three types of data structures: lists, tail queues, and circular queues. All three structures support the following functionality:
> * Insertion of a new entry at the head of the list.
> * Insertion of a new entry after any element in the list.
> * Removal of any entry in the list.
> * Forward traversal through the list. 
>
> Lists are the simplest of the three data structures and support only the above functionality.
>
>
> Tail queues add the following functionality:
> * Entries can be added at the end of a list.
>
> However:
> 1.    All list insertions and removals must specify the head of the list.
> 2.    Each head entry requires two pointers rather than one.
> 3.    Code size is about 15% greater and operations run about 20% slower than lists. 
>
>
> Circular queues add the following functionality:
> * Entries can be added at the end of a list.
> * Entries can be added before another entry.
> * They may be traversed backward, from tail to head. 
>
> However:
> 1. All list insertions and removals must specify the head of the list.
> 2. Each head entry requires two pointers rather than one.
> 3. The termination condition for traversal is more complex.
> 4. Code size is about 40% greater and operations run about 45% slower than lists. 

Zig's standard library should have answers for each of these.
I propose that `std.LinkedList` becomes a union of the three different types.

**Comment by shawnl** [comment]
I wrote up a differn't linked list here:

https://github.com/shawnl/zig/tree/linked-list

**Comment by daurnimator** [comment]
> I wrote up a differn't linked list here:
> 
> https://github.com/shawnl/zig/tree/linked-list

Thanks. A couple of interesting things in there from my perspective:
  - thread safety
  - compat with robust_list API

**Comment by shawnl** [comment]
>  thread safety

No, it is NOT thread safe. That is very difficult with a lockless algorithm.

**Comment by daurnimator** [comment]
> No, it is NOT thread safe. That is very difficult with a lockless algorithm.

Not entirely; but why else do you have all the `@atomicRmw` annotated operations?
